kopf/clients/auth.py

```diff
@@ -1,42 +1,125 @@
+    Login to Kubernetes cluster, locally or remotely.
+
+    Keep the logged in state or config object in the global variables,
+    so that it can be available for future calls via the same function call.
+
+    Automatic refresh/reload of the tokens or objects also should be done here.
     """
 
-    # Configure the default client credentials for all possible environments.
+    # Pykube login is mandatory. If it fails, the framework will not run at all.
+    try:
+        import pykube
+    except ImportError:
+        raise  # mandatory
+    else:
+        login_pykube(verify=verify)
+
+    # We keep the official client library auto-login only because it was
+    # an implied behavior before switching to pykube -- to keep it so (implied).
+    try:
+        import kubernetes
+    except ImportError:
+        pass  # optional
+    else:
+        login_client(verify=verify)
+
+
+def login_pykube(verify=False):
+    global _pykube_cfg
+    try:
+        _pykube_cfg = pykube.KubeConfig.from_service_account()
+        logger.debug("Pykube is configured in cluster with service account.")
+    except FileNotFoundError:
+        try:
+            _pykube_cfg = pykube.KubeConfig.from_file()
+            logger.debug("Pykube is configured via kubeconfig file.")
+        except (pykube.PyKubeError, FileNotFoundError):
+            raise LoginError(f"Cannot authenticate pykube neither in-cluster, nor via kubeconfig.")
+
+    if verify:
+        verify_pykube()
+
+
+def login_client(verify=False):
+    import kubernetes.client
     try:
         kubernetes.config.load_incluster_config()  # cluster env vars
-        logger.debug("configured in cluster with service account")
+        logger.debug("Client is configured in cluster with service account.")
     except kubernetes.config.ConfigException as e1:
         try:
             kubernetes.config.load_kube_config()  # developer's config files
-            logger.debug("configured via kubeconfig file")
+            logger.debug("Client is configured via kubeconfig file.")
         except kubernetes.config.ConfigException as e2:
-            raise LoginError(f"Cannot authenticate neither in-cluster, nor via kubeconfig.")
+            raise LoginError(f"Cannot authenticate client neither in-cluster, nor via kubeconfig.")
 
-    # Make a sample API call to ensure the login is successful,
-    # and convert some of the known exceptions to the CLI hints.
+    if verify:
+        verify_client()
+
+
+def verify_pykube():
+    """
+    Verify if login has succeeded, and the access configuration is still valid.
+    """
+    try:
+        api = get_pykube_api()
+        rsp = api.get(version="", base="/")
+        rsp.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        if e.response.status_code == 401:
+            raise AccessError("Cannot authenticate to the Kubernetes API. "
+                              "Please login or configure the tokens.")
+        else:
+            raise
```

Comment:
I was surprised to get a `403` from there the other day. I eventually discovered it was because my operator's service account was missing the following role:
```yaml
nonResourceURLs: ["/"]
verbs: ["get"]
```
(This was never a problem when Kopf was using the `kubernetes` client.)
Do you think it would be worth catching and mentioning here (not necessarily in this PR)?

Reply:
Good catch, I think requesting `/` might be tricky to not forget in RBAC setups

Reply:
Good point. Maybe, there should be no verification at all. It should fail on the actual operations attempted. 

Looking wider, it can be so that authentication succeeds, but the authorization fails on specific resources or specific REST operations of those resources, such as `PATCH` (because of RBAC). The operator should perform a full self-check on startup. It will only cover the permissions needed by the framework (watching-patching), but will not cover the permissions needed by the operator's logic (e.g. creating pods/jobs). 

See also: #49 for such self-checks (against a RBAC yaml, not against the real cluster, but it can be extended).

Meanwhile, I will make it here to ignore all errors except for HTTP 401 and TCP/HTTPS connection errors — since the purpose is to verify authentication, not authorization.

Reply:
Done in 2118b8ed0f9f06822ba694a4751784804267d8dc. Now, only connection errors and 401. All other errors are ignored.